Autopilot/PathManager/Src/pathStateClasses.cpp

```diff
@@ -175,7 +330,360 @@ pathManagerState& fatalFailureMode::getInstance()
+    {
+        pathMgr->stage = TOUCHDOWN;
+    }
+    else
+    {
+        //align heading with landing direction
+        output.desiredHeading = input.telemetryData->stoppingDirectionHeading;
+        output.useHeading = true;
+        
+        //retrieving desired slow flight speed
+        output.desiredAirspeed = LandingTakeoffManager::slowFlightSpeed(pathMgr->isPackage);
+        
+        //throttle off
+        output.controlDetails.throttlePassby = true;
+        output.controlDetails.throttlePercent = 0;
+    }
+
+    if(pathMgr->isError)
+    {
+        pathMgr->setState(fatalFailureMode::getInstance());
+    }
+    else
+    {
+        pathMgr->setState(coordinateTurnElevation::getInstance());
+    }
+}
+
+pathManagerState& landingDecrabStage::getInstance()
+{
+    static landingDecrabStage singleton;
+    return singleton;
+}
+
+void landingTouchdownStage::execute(pathManager* pathMgr)
+{
+    //load in sensor fusion data and telemtry data into input structure
+    input.telemetryData = commsWithTelemetry::GetTelemetryIncomingData();
+    input.sensorOutput = sensorFusion::GetSFOutput();
+
+    //throttle off
+    output.controlDetails.throttlePassby = true;
+    output.controlDetails.throttlePercent = 0;
+
+    //aligning heading
+    output.desiredHeading = input.telemetryData->stoppingDirectionHeading;
+    output.useHeading = true;
+
+    //ensure made landing points is reset
+    pathMgr->madeLandingPoints = false;
+
+    if(pathMgr->isError)
+    {
+        pathMgr->setState(fatalFailureMode::getInstance());
+    }
+    else
+    {
+        pathMgr->setState(coordinateTurnElevation::getInstance());
+    }
+}
+
+pathManagerState& landingTouchdownStage::getInstance()
+{
+    static landingTouchdownStage singleton;
+    return singleton;
+}
 
 
+/****************************************************************************************************
+TAKEOFF STATE FUNCTIONS
+****************************************************************************************************/
 
+void takeoffRollStage::execute(pathManager* pathMgr)
+{
+    //load in sensor fusion data and telemtry data into input structure
+    input.telemetryData = commsWithTelemetry::GetTelemetryIncomingData();
+    input.sensorOutput = sensorFusion::GetSFOutput();
+
+    //max throttle
+    output.controlDetails.throttlePassby = true;
+    output.controlDetails.throttlePercent = 100;
+
+    if(!pathMgr->madeTakeoffPoints)
+    {
+        takeoffPoint = LandingTakeoffManager::createTakeoffWaypoint(input.sensorOutput->latitude,input.sensorOutput->longitude, input.sensorOutput->altitude, input.telemetryData->takeoffDirectionHeading);
+        pathArray[0] = takeoffPath.initialize_waypoint(takeoffPoint.longitude, takeoffPoint.latitude, takeoffPoint.altitude, PATH_FOLLOW);
+        currentLocation = takeoffPath.initialize_waypoint(input.sensorOutput->longitude, input.sensorOutput->latitude, input.sensorOutput->altitude, HOLD_WAYPOINT);
```

Comment:
I'd recommend adding 10-15 metres in the altitude of currentLocation so it's not at ground level